src/pycell/impl_.rs

```diff
@@ -50,22 +51,43 @@ impl<M: PyClassMutability> PyClassMutability for ExtendsMutableAncestor<M> {
     type MutableChild = ExtendsMutableAncestor<MutableClass>;
 }
 
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
-struct BorrowFlag(usize);
+#[derive(Debug)]
+struct BorrowFlag(AtomicUsize);
 
 impl BorrowFlag {
-    pub(crate) const UNUSED: BorrowFlag = BorrowFlag(0);
-    const HAS_MUTABLE_BORROW: BorrowFlag = BorrowFlag(usize::MAX);
-    const fn increment(self) -> Self {
-        Self(self.0 + 1)
+    pub(crate) const UNUSED: usize = 0;
+    const HAS_MUTABLE_BORROW: usize = usize::MAX;
+    fn increment(&self) -> Result<(), PyBorrowError> {
+        let value = self.0.load(Ordering::Relaxed);
+        if value == BorrowFlag::HAS_MUTABLE_BORROW {
+            return Err(PyBorrowError { _private: () });
+        }
+        loop {
+            match self
+                .0
+                .compare_exchange(value, value + 1, Ordering::SeqCst, Ordering::SeqCst)
+            {
+                Ok(..) => {
+                    // value successfully incremented
+                    break Ok(());
+                }
+                Err(..) => {
+                    // value changed under us, need to reload and try again
+                    let value = self.0.load(Ordering::Relaxed);
```

Comment:
Do we need to reload? Isn't the value in `Err(..)` what we want?

Reply:
Oh, either way: the `let` here causes `value` to be rebound, which means that teh compare_exchange is always comparing to the original value, which I don't think is what you want.

Reply:
thanks! much clearer this way without the second load.